### DISCUSSION TITLE
Typing issue with __table__ and __mapper__

### DISCUSSION THREAD

**Discussion by prousso** [sqlalchemy/sqlalchemy#11629] [Usage Questions] [answered]
Hi, I have some typing issue with following code:
```python
import uuid

from sqlalchemy import Column
from sqlalchemy.dialects import postgresql

from sqlalchemy.orm import registry

mapper_registry = registry(constructor=lambda x: None)


@mapper_registry.as_declarative_base()
class Base:
    pass


class SqlItem(Base):
    __tablename__ = "cactus"

    id = Column(postgresql.UUID(as_uuid=True), primary_key=True)


SqlItem.__table__.columns
```
mypy:  error: "type[SqlItem]" has no attribute "__table__"
I have same issue with __mapper__

python:3.12
sqlalchemy: 2.0.31
mypy: 1.10.1

Thansk

**Comment by CaselIT** [comment]
Hi

As in the documentation you need to subclass `DeclarativeBase` instead of `mapper_registry.as_declarative_base()`

To manually specify a registry you can assign to it in the base

**Reply by prousso** [reply]
Sorry I forgot but I have metadata, if I use `DeclarativeMeta` I will have the same issue (with extra `Missing positional arguments "classname", "bases", "dict_" in call to "SqlItem"`)

**Reply by CaselIT** [reply]
I don't follow. Can you provide an example?

**Reply by prousso** [reply]
I think I found my issue.
My code was:
```python
import uuid
from uuid import uuid4

from sqlalchemy import Column, MetaData
from sqlalchemy.dialects import postgresql

from sqlalchemy.orm import DeclarativeMeta

meta = MetaData(
    naming_convention={
        "ix": "ix_%(column_0_label)s",
    }
)


class Base(DeclarativeMeta):
    metadata = meta

class SqlItem(Base):
    __tablename__ = "cactus"

    id = Column(postgresql.UUID(as_uuid=True), primary_key=True)


SqlItem.__table__.columns
item = SqlItem(id=uuid4())

```

But I think it should be:
```python
import uuid
from uuid import uuid4

from sqlalchemy import Column, MetaData
from sqlalchemy.dialects import postgresql

from sqlalchemy.orm import DeclarativeBase

meta = MetaData(
    naming_convention={
        "ix": "ix_%(column_0_label)s",
    }
)


class Base(DeclarativeBase):
    metadata = meta

class SqlItem(Base):
    __tablename__ = "cactus"

    id = Column(postgresql.UUID(as_uuid=True), primary_key=True)


SqlItem.__table__.columns
item = SqlItem(id=uuid4())

```

**Reply by CaselIT** [reply, marked as answer]
yes, `DeclarativeMeta` is a metaclass, that can be used to create custom instrumentations.

Also note that if you want to use typing using Column and `id = Column` will not really work. You need to follow the documentation on how to define the models to make type checker happy. https://docs.sqlalchemy.org/en/20/changelog/whatsnew_20.html#whatsnew-20-orm-declarative-typing

**Reply by prousso** [reply]
Yes of course
